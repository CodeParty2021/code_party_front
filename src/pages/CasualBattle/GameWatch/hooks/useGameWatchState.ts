import { useEffect } from "react";
import { useNavigate } from "react-router-dom";

import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import {
  ResultType,
  useFetchResult,
} from "hooks/ResultAPIHooks/useFetchResult";
import { useFetchJson } from "hooks/ResultAPIHooks/useFetchJson";

export type IResponse = {
  isAnalyzing: boolean;
  analyzingError: boolean;
  result?: ResultType;
  json?: string;
  exitBtnHandler: () => void;
};

export const useGameWatchState = (): IResponse => {
  const { room, updateMember } = useRoomSync();
  const { data: result, fetchResult } = useFetchResult();
  const { data: json, fetchJson } = useFetchJson();
  const navigate = useNavigate();

  //初期処理
  useEffect(() => {
    //ユーザ状態を更新
    updateMember({
      status: "watching",
    });
  }, []);

  // 何らかの理由でルームから出された場合はロビーに戻る
  useEffect(() => {
    if (!room.isEntered) {
      navigate("/casual-battle");
    }
  }, [room.isEntered]);

  //シミュレーション結果が更新されたら結果をフェッチする
  useEffect(() => {
    if (room.info?.analyzingResult?.resultId) {
      fetchResult(room.info?.analyzingResult.resultId);
      fetchJson(room.info?.analyzingResult.resultId);
    }
  }, [room.info?.analyzingResult]);

  const _exitBtnHandler = () => {
    navigate("/casual-battle/waiting-room");
  };

  return {
    isAnalyzing: !room.info?.analyzingResult?.resultId,
    analyzingError: Boolean(room.info?.analyzingResult?.error),
    result: result,
    json: json,
    exitBtnHandler: _exitBtnHandler,
  };
};
