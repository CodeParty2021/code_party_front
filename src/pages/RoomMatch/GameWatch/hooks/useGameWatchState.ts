import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";

import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { useFetchJson } from "hooks/ResultAPIHooks/useFetchJson";
import { useUnityGame } from "hooks/UnityGameHooks/useUnityGame";
import { UnityContext } from "react-unity-webgl";
import { useDummyLoading } from "hooks/DummyLoadingHooks/useDummyLoading";

export type IResponse = {
  unityContext: UnityContext;
  state: "Analyzing" | "Running" | "Finish" | "AnalyzingError";
  messageType: "Analyzing" | "AnalyzingError";
  exitBtnHandler: () => void;
};

export const useGameWatchState = (): IResponse => {
  const { room, updateMember } = useRoomSync();
  const { data: json, fetchJson } = useFetchJson();
  const { unityContext, startGame, unityStatus } = useUnityGame("SquarePaint");
  const [messageType, setMessageType] =
    useState<IResponse["messageType"]>("Analyzing");
  const { dummyLoadingState, startDummyLoad } = useDummyLoading(1000);
  const navigate = useNavigate();

  const state = room.info?.analyzingResult?.error
    ? "AnalyzingError"
    : !room.info?.analyzingResult?.resultId ||
      unityStatus.isLoading ||
      dummyLoadingState.isLoading
    ? "Analyzing"
    : unityStatus.isRunning
    ? "Running"
    : "Finish";

  // メッセージの操作
  useEffect(() => {
    if (state === "Analyzing") setMessageType("Analyzing");
    else if (state === "AnalyzingError") setMessageType("AnalyzingError");
  }, [state]);

  //初期処理
  useEffect(() => {
    //ユーザ状態を更新
    updateMember({
      status: "watching",
    });

    // ダミーロード開始
    startDummyLoad();
  }, []);

  // 何らかの理由でルームから出された場合はロビーに戻る
  useEffect(() => {
    if (!room.isEntered) {
      navigate("/room-match");
    }
  }, [room.isEntered]);

  //シミュレーション結果が更新されたら結果をフェッチする
  useEffect(() => {
    if (room.info?.analyzingResult?.resultId) {
      fetchJson(room.info?.analyzingResult.resultId);
    }
  }, [room.info?.analyzingResult]);

  // 観戦開始
  useEffect(() => {
    if (!unityStatus.isLoading && json) {
      startGame(json);
    }
  }, [unityStatus.isLoading, json]);

  const _exitBtnHandler = () => {
    navigate("/room-match/waiting-room");
  };

  return {
    state,
    messageType,
    unityContext,
    exitBtnHandler: _exitBtnHandler,
  };
};
