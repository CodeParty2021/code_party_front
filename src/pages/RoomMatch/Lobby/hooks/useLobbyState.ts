import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { RefObject, useCallback, useEffect, useMemo, useRef, useState } from "react";
import { useNavigate } from "react-router-dom";

export type IResponse = {
  isLoading: boolean;
  errorMessage: string;
  roomIdRef: RefObject<HTMLInputElement>;
  createRoomHandler: () => void;
  createRoomDisabled: boolean;
  enterRoomHandler: () => Promise<void>;
  enterRoomDisabled: boolean;
  backButtonHandler: () => void;
};

export const useLobbyState = (): IResponse => {
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const [errorMessage, setErrorMessage] = useState<string>("");
  const [isProcessingAny, setIsProcessingAny] = useState<boolean>(false);
  const [succeedEnterRoom, setSucceedEnterRoom] = useState<boolean>(false);
  const { room, createRoom, enterRoom } = useRoomSync();
  const navigate = useNavigate();
  const roomIdRef = useRef<HTMLInputElement>(null!);

  const createEnterButtonDisabled = useMemo(() => (
    isProcessingAny || room.isEntered || succeedEnterRoom
  ), [isProcessingAny, room.isEntered]);

  useEffect(() => {
    if (room.isEntered) {
      navigate("/room-match/waiting-room");
    }
  }, [room.isEntered]);

  // ルーム
  const _createRoomHandler = useCallback(async () => {
    setIsLoading(true);
    setIsProcessingAny(true);
    await createRoom()
      .then(() => {
        setSucceedEnterRoom(true);
      })
      .catch(() => {
        navigate("/error");
      });
    setIsProcessingAny(false);
  }, []);

  const _enterRoomHandler = useCallback(async () => {
    const value = roomIdRef.current?.value;

    setIsProcessingAny(true);
    if (typeof value === "string") {
      if (value == "") {
        setErrorMessage("値を入力してください。");
      } else {
        await new Promise((resolve) => setTimeout(resolve, 500));
        await enterRoom(value).then(() => {
          setSucceedEnterRoom(true);
        }).catch((e) => {
          if (e.message == "roomId is empty") {
            setErrorMessage("値を入力してください。");
          } else if (e.message == "roomId is not found") {
            setErrorMessage("ルームIDが無効です。");
          } else if (e.message == "room is full") {
            setErrorMessage("ルームが満員です。");
          }
        });
      }
    } else {
      setErrorMessage("入力が不正です。");
    }
    setIsProcessingAny(false);
  }, []);

  // 戻るボタン
  const _backButtonHandler = useCallback(() => {
    navigate("/select-mode");
  }, []);

  return {
    isLoading,
    errorMessage,
    roomIdRef,
    createRoomHandler: _createRoomHandler,
    createRoomDisabled: createEnterButtonDisabled,
    enterRoomHandler: _enterRoomHandler,
    enterRoomDisabled: createEnterButtonDisabled,
    backButtonHandler: _backButtonHandler,
  };
};
