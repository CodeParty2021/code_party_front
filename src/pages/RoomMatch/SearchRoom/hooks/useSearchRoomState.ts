import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";

import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";

export type IResponse = {
  roomIdTextBoxValue: string;
  roomIdTextBoxChangeHandler: (value: string) => void;
  enterBtnDisabled: boolean;
  enterBtnClickHandler: () => void;
};

export const useSearchRoomState = (): IResponse => {
  const { room, enterRoom } = useRoomSync();
  const [roomId, setRoomId] = useState("");
  const [enterBtnDisabled, setEnterBtnDisabled] = useState(false);
  const navigate = useNavigate();

  useEffect(() => {
    if (room.isEntered) {
      navigate("/room-match/waiting-room");
    }
  }, [room.isEntered]);

  const _roomIdTextBoxChangeHandler = (value: string) => {
    setRoomId(value);
  };

  const _enterBtnHandler = () => {
    setEnterBtnDisabled(true);
    enterRoom(roomId).catch(() => {
      setEnterBtnDisabled(false);
    });
  };

  return {
    roomIdTextBoxValue: roomId,
    roomIdTextBoxChangeHandler: _roomIdTextBoxChangeHandler,
    enterBtnDisabled: enterBtnDisabled,
    enterBtnClickHandler: _enterBtnHandler,
  };
};
