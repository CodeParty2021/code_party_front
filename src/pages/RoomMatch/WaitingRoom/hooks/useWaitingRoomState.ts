import { useEffect, useState, useMemo } from "react";
import { useNavigate, useSearchParams } from "react-router-dom";

import { useRoomSync } from "hooks/RoomSyncHooks/useRoomSync";
import { CodeType, useFetchCodes } from "hooks/CodeAPIHooks/useFetchCodes";
import { UserState, UserAction } from "services/RoomSync/RoomSync";
import { RootState } from "store";
import { useSelector } from "react-redux";
import { User } from "services/user/user";

export type IResponse = {
  user: User | null;
  roomInfo: {
    roomId: string;
    invitationLink: string;
    host: UserState;
    memberKeys: string[];
    members: { [id: string]: UserState };
    actionKeys: string[];
    actions: { [id: string]: UserAction };
  };
  isHost: boolean;
  status?: "waiting" | "watching";
  ready: boolean;

  readyBtnHandler: () => void;
  readyBtnDisabled: boolean;
  exitBtnHandler: () => void;
  startBtnDisabled: boolean;
  startBtnHandler: () => void;
  isCopyBtnClicked: boolean;
  invitationBtnHandler: () => void;
  kickUserHandler: (userId: string) => void;
  code: {
    codes: CodeType[];
    loading: boolean;
  };
  selectedCode: CodeType | null;
  onChangeSelectedCode: (code: CodeType) => void;
  onStartToEditCode: () => void;
  showCodeSelectModal: boolean;
  openCodeSelectModal: () => void;
  closeCodeSelectModal: () => void;
};

export const useWaitingRoomState = (): IResponse => {
  const {
    room,
    isHost,
    updateRoomInfo,
    updateMember,
    updateOtherMember,
    exitRoom,
  } = useRoomSync();
  const { data: codes = [], loading } = useFetchCodes();
  const [preRoomStatus, setPreRoomStatus] = useState<"waiting" | "watching">(
    "watching"
  );
  const [selectedCode, setSelectedCode] = useState<CodeType | null>(null);
  const [ready, setReady] = useState(false);
  const [readyBtnDisabled, setReadyBtnDisabled] = useState(true);
  const [startBtnDisabled, setStartBtnDisabled] = useState(true);
  const [isCopyBtnClicked, setIsCopyBtnClicked] = useState(false);
  let { user } = useSelector((state: RootState) => state.user);
  const navigate = useNavigate();
  const [searchParams] = useSearchParams();
  const dummyUser: UserState = {
    displayName: "",
    status: "disconnect",
    ready: false,
  };

  const modalOn = useMemo(
    () => (searchParams.get("modal") === "on" ? true : false),
    []
  );

  // ユーザ状態を更新
  useEffect(() => {
    updateMember({
      status: "waiting",
    });
  }, []);

  // DB上で退出処理が完了したら退出
  useEffect(() => {
    if (!room.isEntered) {
      navigate("/room-match");
    }
  }, [room.isEntered]);

  // 準備ボタンの有効化
  useEffect(() => {
    //コードが選択されたら準備完了ボタンを押せるようにする
    if (selectedCode != null) {
      setReadyBtnDisabled(false);
    }
  }, [selectedCode]);

  // ルームステータスがwaitingからwatchingになったら観戦画面に遷移
  useEffect(() => {
    if (room.info) {
      if (preRoomStatus == "waiting" && room.info.status == "watching") {
        navigate("/room-match/result");
      }
      setPreRoomStatus(room.info?.status);
    }
  }, [room.info?.status]);

  // ready状態をDBに反映
  useEffect(() => {
    updateMember({
      ready: ready,
      codeId: selectedCode ? selectedCode.id : "",
    });
  }, [ready]);

  // クエリパラメータにmodal=onを渡されたらモーダルを表示
  useEffect(() => {
    if (modalOn) {
      setShowCodeSelectModal(true);
    }
  }, [modalOn]);

  // ホストの処理
  useEffect(() => {
    if (isHost) {
      if (room.info && room.info.status == "waiting") {
        // 全員が準備完了になったら開始ボタンを有効にする
        let newDisabled = false;
        for (let key of Object.keys(room.members)) {
          if (!room.members[key].ready) {
            newDisabled = true;
            break;
          }
        }
        setStartBtnDisabled(newDisabled);
      }
      // 全員がWaitingRoomに返ってきたらroomの状態をwaitingに更新する
      let newRoomStatus: "waiting" | "watching" = "waiting";
      for (let key of Object.keys(room.members)) {
        if (room.members[key].status == "watching") {
          newRoomStatus = "watching";
        }
      }
      if (newRoomStatus == "waiting" && room.info?.status == "watching") {
        updateRoomInfo({
          status: "waiting",
          analyzingResult: null,
        });
      }
    }
  }, [room.members, isHost]);

  // kick処理 member.status = kickingの場合そのユーザをキックする
  useEffect(() => {
    if (user) {
      for (let key of Object.keys(room.members)) {
        if (room.members[key].status == "kicking" && user.id == key) {
          exitRoom();
        }
      }
    }
  }, [room.members]);

  // ---- イベントハンドラ ---- //

  const _readyBtnHandler = () => {
    setReady(!ready);
  };

  const _exitBtnHandler = () => {
    exitRoom();
  };

  const _startBtnHandler = () => {
    updateRoomInfo({
      status: "watching",
    });
  };

  const _onChangeSelectedCode = (code: CodeType) => {
    setSelectedCode(code);
  };

  const _onStartToEditCode = () => {
    if (selectedCode) {
      navigate(`/free-coding/${selectedCode.id}/selectCode`);
    }
  };

  //invitationURLのコピーボタン
  const _invitationBtnHandler = () => {
    if (room.invitationLink) {
      navigator.clipboard.writeText(room.invitationLink).then(() => {
        setIsCopyBtnClicked(true);
      });
    }
  };
  const _kickUserBtnHandler = (userId: string) => {
    updateOtherMember(userId, { status: "kicking" });
  };

  //モーダル表示非表示
  const [showCodeSelectModal, setShowCodeSelectModal] =
    useState<boolean>(false);
  const openCodeSelectModal = () => {
    setShowCodeSelectModal(true);
  };
  const closeCodeSelectModal = () => {
    setShowCodeSelectModal(false);
  };

  return {
    user: user,
    roomInfo: {
      roomId: room.id ? room.id : "",
      invitationLink: room.invitationLink ? room.invitationLink : "",
      host:
        room.info && room.info.host in room.members
          ? room.members[room.info.host]
          : dummyUser,
      memberKeys: room.sortedKeysOfMembers,
      members: room.members,
      actionKeys: room.sortedKeysOfActions,
      actions: room.actions,
    },
    isHost: isHost,
    status: room.info?.status,
    ready: ready,
    readyBtnHandler: _readyBtnHandler,
    readyBtnDisabled: readyBtnDisabled,
    exitBtnHandler: _exitBtnHandler,
    startBtnDisabled: startBtnDisabled,
    startBtnHandler: _startBtnHandler,
    isCopyBtnClicked: isCopyBtnClicked,
    invitationBtnHandler: _invitationBtnHandler,
    kickUserHandler: _kickUserBtnHandler,
    code: {
      codes: codes,
      loading: loading,
    },
    selectedCode: selectedCode,
    onChangeSelectedCode: _onChangeSelectedCode,
    onStartToEditCode: _onStartToEditCode,
    showCodeSelectModal,
    openCodeSelectModal,
    closeCodeSelectModal,
  };
};
