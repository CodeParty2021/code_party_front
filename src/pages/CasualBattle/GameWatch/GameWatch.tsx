import React from "react";
import { useGameWatchState } from "./hooks/useGameWatchState";

type Props = {};

export const CasualBattleGameWatch: React.FC<Props> = () => {
  const {isAnalysing, exitBtnHandler} = useGameWatchState();
  return (
    <div>
      <div>結果</div>
      <div>
        {isAnalysing ? "解析中" : "観戦中"}
      </div>
      <button onClick={exitBtnHandler}>退出</button>
    </div>
  );
};
