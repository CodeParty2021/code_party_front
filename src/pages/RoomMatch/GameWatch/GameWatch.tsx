import React from "react";
import { useGameWatchState } from "./hooks/useGameWatchState";
import { useRunSimulation } from "./hooks/useRunSimulation";
import { Background } from "pages/Code/Coding/CodingStyle";
import { BackLink } from "components/BackLink/BackLink";
import {
  GameWatchStyle,
  messageProps,
  UnityStyle,
  WatchingLogo,
} from "./GameWatchStyle";
import { Message } from "components/Message/Message";
import { AlgoHead } from "components/Character/Algo/Head/AlgoHead";

type Props = {};

export const RoomMatchGameWatch: React.FC<Props> = () => {
  useRunSimulation();
  const { unityContext, exitBtnHandler, state, messageType } =
    useGameWatchState();

  return (
    <GameWatchStyle state={state}>
      <Background color="blue" />
      <div className="gamewatch_unity">
        <UnityStyle unityContext={unityContext} />
        <WatchingLogo src="/img/watching_logo.svg" wrapper="svg" />
      </div>
      <BackLink
        iconColor="blue"
        backMessage="待機部屋に戻る"
        onClick={exitBtnHandler}
      />
      <div className="gamewatch_message">
        <Message {...messageProps(messageType)} />
        <AlgoHead width="142px" height="117px" color="blue" />
      </div>
    </GameWatchStyle>
  );
};
