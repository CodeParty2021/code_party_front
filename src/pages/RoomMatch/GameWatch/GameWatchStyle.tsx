import { Message } from "components/Message/Message";
import { ComponentProps } from "react";
import { ReactSVG } from "react-svg";
import styled, { css } from "styled-components";
import { WHITE } from "styles/colors";
import { IResponse } from "./hooks/useGameWatchState";
import { BackLinkButton } from "components/BackLink/BackLinkStyle";
import Unity from "react-unity-webgl";
import { FlexGap } from "styles/FlexGap/FlexGap";

type PrivateProps = {
  state: IResponse["state"];
};

const defaultStyle = css`
  position: relative;
  width: 100vw;
  height: 100vh;

  /* BackLinkStyle */
  ${BackLinkButton} {
    position: absolute;
    top: 32px;
    left: 32px;
  }

  .gamewatch_message {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    ${FlexGap({ gap: "8px", direction: "column" })}

    position: absolute;
    top: 0;
    left: 50%;
    transform: translate(-50%, -100%);
    transition: all 0.5s ease;
  }

  .gamewatch_unity {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    padding: 12px 48px 0;
    ${FlexGap({ gap: "8px", direction: "column" })}
    transition: opacity 0.5s ease;
  }
`;

const analyzingStyle = css`
  .gamewatch_message {
    top: 50%;
    transform: translate(-50%, calc(-50% - 110px));
  }

  .gamewatch_unity {
    opacity: 0;
  }
`;

const runningStyle = css``;

const finishStyle = css``;

export const GameWatchStyle = styled.div<PrivateProps>`
  ${defaultStyle}

  ${({ state }) => state === "Analyzing" && analyzingStyle}
  ${({ state }) => state === "Running" && runningStyle}
  ${({ state }) => state === "Finish" && finishStyle}
  ${({ state }) => state === "AnalyzingError" && analyzingStyle}
`;

export const UnityStyle = styled(Unity)`
  width: calc(90vh * 4 / 3);
  height: 90vh;
  border-radius: 12px;
  flex-grow: 0;
  flex-shrink: 0;
`;

export const WatchingLogo = styled(ReactSVG)`
  width: 50px;
  height: 32px;
  fill: ${WHITE};
`;

export const messageProps = (
  state: IResponse["state"]
): ComponentProps<typeof Message> => {
  if (state === "Analyzing")
    return {
      title: "シミュレーションチュウ",
      value: "しばらくお待ちください",
      color: "blue",
    };
  else if (state === "AnalyzingError")
    return {
      title: "エラー",
      value: "左上の戻るボタンをおしてください！",
      color: "red",
    };
  else
    return {
      title: "このメッセージは",
      value: "出ないはずだよ",
      color: "blue",
    };
};
