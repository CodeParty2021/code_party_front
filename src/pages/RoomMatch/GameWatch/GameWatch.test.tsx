import React from "react";
import { shallow } from "enzyme";

import { RoomMatchGameWatch } from "./GameWatch";
import { IResponse, useGameWatchState } from "./hooks/useGameWatchState";

jest.mock("./hooks/useGameWatchState");
jest.mock("./hooks/useRunSimulation");

const useGameWatchStateMock = useGameWatchState as jest.Mock;

const state: IResponse = {
  isAnalyzing: false,
  analyzingError: false,
  result: {
    id: "resultId",
    jsonPath: "/tmp/resultId",
    step: 15,
    codes: ["codeid1", "codeid2", "codeid3"],
  },
  exitBtnHandler: jest.fn(),
};

describe("<RoomMatchGameWatch />", () => {
  beforeEach(() => {
    useGameWatchStateMock.mockReturnValue({
      ...state,
    });
  });
  afterEach(() => {
    jest.resetAllMocks();
  });

  it("snapshot test", () => {
    const wrapper = shallow(<RoomMatchGameWatch />);

    expect(wrapper.getElements()).toMatchSnapshot();
  });

  it("exit button test", () => {
    const wrapper = shallow(<RoomMatchGameWatch />);

    const btn = wrapper.find("#exit-btn");
    btn.simulate("click");

    expect(state.exitBtnHandler).toHaveBeenCalled();
  });
});
