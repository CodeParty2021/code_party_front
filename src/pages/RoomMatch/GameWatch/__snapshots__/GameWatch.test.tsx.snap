// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RoomMatchGameWatch /> snapshot test 1`] = `
Array [
  <div>
    <div>
      結果
    </div>
    <div>
      ステータス：
      観戦中
    </div>
    <div>
      {"id":"resultId","jsonPath":"/tmp/resultId","step":15,"codes":["codeid1","codeid2","codeid3"]}
    </div>
    <button
      id="exit-btn"
      onClick={[MockFunction]}
    >
      退出
    </button>
  </div>,
]
`;
